### PR TITLE
Make SuppressInstrumentation an IDisposable

### PR DESCRIPTION
**Pull request by alanwest**
Implementing the suggestion from @reyang and 👍'ed by @CodeBlanch and myself in #960 to introduce `using` syntactic sugar 🍬 for `SuppressInstrumentation`.

Before I go too far, I wanted to have a quick discussion for what kind of behavior we'd like in the event of a nested situation like the following where Dispose is explicitly called:

```csharp
        [Fact]
        public static void UsingSuppressInstrumentation_Nested()
        {
            Assert.False(SuppressInstrumentation.IsSuppressed);
            using (var scope = SuppressInstrumentation.Begin())
            {
                Assert.True(SuppressInstrumentation.IsSuppressed);
                using (var innerScope = SuppressInstrumentation.Begin())
                {
                    Assert.True(SuppressInstrumentation.IsSuppressed);

                    // In order for the next Assert to pass, this Dispose would have to NoOp and be deferred
                    scope.Dispose();
                    Assert.True(SuppressInstrumentation.IsSuppressed);

                    // If we deferred the parent scope's Dispose from above, then it would also need to
                    // be triggered for a dispose here for the next Assert to pass.
                    innerScope.Dispose();
                    Assert.False(SuppressInstrumentation.IsSuppressed);
                }
                Assert.False(SuppressInstrumentation.IsSuppressed);
            }
        }
```

The current logic is simplistic and would not pass this test. Though, if people agree this is the correct flow we'd like to expect, then I can implement it with a mechanism where the inner scope signals the parent that it is ok to dispose. However, I just want to make sure things aren't becoming more complicated than people would hope for.

### CODE REVIEW

**Comment by reyang** [comment]
This is a bit different from what I was originally thinking - I was exploring something on the `RuntimeContext` instead of `SuppressInstrumentation`.

So instead of having:

```csharp
var foo = RuntimeContext.GetValue<int>("foo");
var bar = RuntimeContext.GetValue<string>("bar");
try
{
    RuntimeContext.SetValue("foo", 1);
    RuntimeContext.SetValue("bar", "hello");
    // do something
}
finally
{
    RuntimeContext.SetValue("foo", foo);
    RuntimeContext.SetValue("bar", bar);
}
```

One could use (not sure if people would like `With` or `WithValue`):

```csharp
using (RuntimeContext.With("foo", 1).With("bar", "hello"))
{
    // do something
}
```

And for the flag:

```csharp
using (Sdk.SuppressInstrumentation(true))
{
    // do something
}
```

**Comment by alanwest** [comment]
Hmm, I see. These are interesting ideas. Though both of them still raise the question of what the behavior should be if a `using` is used _and_ `Dispose` is explicitly called prior to the end of the `using` scope. I might be overthinking things (I do that) - though just looking to be defensive. I'm currently looking to see if there is a common practice here. Like in logging frameworks where nested scopes are common.

**Comment by CodeBlanch** [comment]
I kind of like what @alanwest has. I would probably add "Scope" (or "Context") into the name, a la:
```csharp
using (SuppressInstrumentationScope.Begin())
{
   await ExportAsync(batch).ConfigureAwait(false);
}
```
I think for the common case we don't need to pass `true`? Meaning, I can't think of a use case for setting it as false.

Now if SuppressInstrumentationScope was just wrapping `RuntimeContext.With("otel.suppress_instrumentation", true)` that would be really neat.

> // In order for the next Assert to pass, this Dispose would have to NoOp and be deferred

What you are doing is (essentially):
```csharp
            using (var scope = SuppressInstrumentation.Begin())
                using (var innerScope = SuppressInstrumentation.Begin())
                    scope.Dispose();
                    innerScope.Dispose();
```
I would simply say, unsupported 😄 Have to dispose them in the correct order.
```csharp
            using (var scope = SuppressInstrumentation.Begin())
                using (var innerScope = SuppressInstrumentation.Begin())
                    innerScope.Dispose();
                    scope.Dispose();
```
That should work fine.

**Comment by alanwest** [comment]
I do like the `unsupported 😄` approach. Can we put emojis in the documentation?

> I think for the common case we don't need to pass true? Meaning, I can't think of a use case for setting it as false.

I was also imagining this to be true. Though also trying to stretch my brain in search of use cases I may not be thinking of. @reyang, was there a use case you were thinking of?

**Comment by CodeBlanch** [comment]
> I do like the `unsupported 😄` approach. Can we put emojis in the documentation?

I'm being serious! hah I don't think the expectation on these scope/context type of things is that you can dispose them out of order. I'm thinking about Activity.Current. It has the same issue. If you dispose/stop a parent, it sets its parent as Activity.Current. If someone then does Activity.Current.Stop thinking it's the child, havoc ensues. Same idea for log scopes.

**Comment by reyang** [comment]
> > I think for the common case we don't need to pass true? Meaning, I can't think of a use case for setting it as false.
> 
> I was also imagining this to be true. Though also trying to stretch my brain 🧠 in search of use cases I may not be thinking of. @reyang, was there a use case you were thinking of?

I don't have a use case at all. It was mainly coming from natural language. For example `using(SuppressInstrumentation)` makes sense literally, the same for `using(SuppressInstrumentation(true))`, however `using(SuppressInstrumentation.Begin())` seems weird literally as folks would start to wonder **Where is `SuppressInstrumentation.End()`?**.

**Comment by CodeBlanch** [comment]
It reads really naturally to me, but I'm working in a codebase that uses a lot of [log scopes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.beginscope?view=dotnet-plat-ext-3.1). Or its cousin, the [change token](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.primitives.changetoken.onchange?view=dotnet-plat-ext-3.1). Basically a newer pattern in .NET where you return IDisposable as the mechanism for calling stop/end/cancel. I really like it 🤷

**Comment by reyang** [comment]
> It reads really naturally to me, but I'm working in a codebase that uses a lot of [log scopes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.beginscope?view=dotnet-plat-ext-3.1). Or its cousin, the [change token](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.primitives.changetoken.onchange?view=dotnet-plat-ext-3.1). Basically a newer pattern in .NET where you return IDisposable as the mechanism for calling stop/end/cancel. I really like it 🤷

Then go for it 💯 😄

**Comment by reyang** [comment]
> I do like the `unsupported 😄` approach. Can we put emojis in the documentation?

FYI https://github.com/open-telemetry/opentelemetry-cpp/pull/209#discussion_r460363669.

**Comment by alanwest** [comment]
> > I do like the `unsupported 😄` approach. Can we put emojis in the documentation?
> 
> I'm being serious! hah I don't think the expectation on these scope/context type of things is that you can dispose them out of order.

Yea me, too! Sorry if I sounded sarcastic 😅. I completely agree that it make sense not to support out-of-order disposal. I figured this was a case of over-thinking, so I always like to just stop and see what others think.


> > It reads really naturally to me, but I'm working in a codebase that uses a lot of [log scopes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.beginscope?view=dotnet-plat-ext-3.1). Or its cousin, the [change token](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.primitives.changetoken.onchange?view=dotnet-plat-ext-3.1). Basically a newer pattern in .NET where you return IDisposable as the mechanism for calling stop/end/cancel. I really like it 🤷
> 
> Then go for it 💯 😄

Sounds like a decent consensus here. I appreciate the thoughts around natural language, though I also agree that there is enough of a pattern and practice in .NET for the `Begin()` approach to be familiar with folk.

I'll get this PR cleaned up.

File: src/OpenTelemetry/SuppressInstrumentation.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            this.parent = SuppressInstrumentationRuntimeContextSlot.Get();
+            SuppressInstrumentationRuntimeContextSlot.Set(this);
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether automatic telemetry
+        /// collection in the current context should be suppressed (disabled).
+        /// </summary>
+        public static bool IsSuppressed => SuppressInstrumentationRuntimeContextSlot.Get() != default;
```
Comment (reyang) [inline src/OpenTelemetry/SuppressInstrumentation.cs:39]: Consider implicate operator?

```csharp
public sealed class SuppressInstrumentationScope : IDisposable
{
    private readonly bool value;
    public static implicit operator bool(SuppressInstrumentationScope _) => SuppressInstrumentationRuntimeContextSlot.Get();

    public SuppressInstrumentationScope(bool value = true)
    {
        this.value = value;
        SuppressInstrumentationRuntimeContextSlot.Set(value);
    }

    public void Dispose()
    {
        SuppressInstrumentationRuntimeContextSlot.Set(this.value);
    }

    public IDisposable Begin(bool value = true)
    {
        return new SuppressInstrumentationScope(value);
    }
}

public static SuppressInstrumentationScope SuppressInstrumentation { get; set; } = new SuppressInstrumentationScope(false);
```

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentation.cs:39]: The for consumption:

```csharp
if (Sdk.SuppressInstrumentation)
{
    // do something
}

using (Sdk.SuppressInstrumentation.Begin(true))
{
    // do something
}
```

File: src/OpenTelemetry/SuppressInstrumentation.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        /// <returns>Object to dispose to end the scope.</returns>
+        /// <remarks>
+        /// This is typically used to prevent infinite loops created by
+        /// collection of internal operations, such as exporting traces over HTTP.
+        /// <code>
+        ///    public override async Task&lt;ExportResult&gt; ExportAsync(
+        ///        IEnumerable&lt;Activity&gt; batch,
+        ///        CancellationToken cancellationToken)
+        ///    {
+        ///       using (var scope = SuppressInstrumentation.Begin())
```
Comment (reyang) [inline src/OpenTelemetry/SuppressInstrumentation.cs:53]: nit, remove `scope`.

File: src/OpenTelemetry/SuppressInstrumentation.cs (hunk 1 above)
Comment (alanwest) [reply src/OpenTelemetry/SuppressInstrumentation.cs:39]: Interesting, I would not have thought of this approach!

Comment (alanwest) [reply src/OpenTelemetry/SuppressInstrumentation.cs]: Hmm, this isn't working for some reason. I'll take closer look in the morning.

Basically, I debugged through this code and...
```csharp
using (SuppressInstrumentationScope.Begin()) // This does in fact invoke SuppressInstrumentationRuntimeContextSlot.Set(true);
{
    Assert.True(Sdk.SuppressInstrumentation); // This fails, even though the implicit operator is invoked
}
```

In the meantime I had a few changes staged prior to seeing this idea, so I just pushed them up.

Comment (alanwest) [reply src/OpenTelemetry/SuppressInstrumentation.cs]: Oh.. silly me, the static initializer for `Sdk.SuppressInstrumentation` had not yet been invoked prior to calling `SuppressInstrumentationScope.Begin()`

Comment (alanwest) [reply src/OpenTelemetry/SuppressInstrumentation.cs]: Ok pushed up something that works. Ignore my initial confused comments.

File: src/OpenTelemetry/SuppressInstrumentationScope.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+// </copyright>
+
+using System;
+using OpenTelemetry.Context;
+
+namespace OpenTelemetry
+{
+    public sealed class SuppressInstrumentationScope : IDisposable
+    {
+        private static readonly RuntimeContextSlot<bool> SuppressInstrumentationRuntimeContextSlot = RuntimeContext.RegisterSlot<bool>("otel.suppress_instrumentation");
```
Comment (reyang) [inline src/OpenTelemetry/SuppressInstrumentationScope.cs:24]: Given this is inside a class `SuppressInstrumentationScope`, we can probably simplify the name to `ContextSlot` or even `Slot`.

File: src/OpenTelemetry/SuppressInstrumentationScope.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        private readonly bool previousValue;
+        private bool disposed;
+
+        internal SuppressInstrumentationScope(bool value = true)
+        {
+            this.previousValue = SuppressInstrumentationRuntimeContextSlot.Get();
+            SuppressInstrumentationRuntimeContextSlot.Set(value);
+        }
+
+        public static implicit operator bool(SuppressInstrumentationScope x) => SuppressInstrumentationRuntimeContextSlot.Get();
```
Comment (reyang) [inline src/OpenTelemetry/SuppressInstrumentationScope.cs:35]:
```suggestion
        public static implicit operator bool(SuppressInstrumentationScope unused) => SuppressInstrumentationRuntimeContextSlot.Get();
```

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:35]: Or other name that explains the intention.

File: src/OpenTelemetry/SuppressInstrumentationScope.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+        /// <summary>
+        /// Begins a new scope in which automatic telemetry is suppressed (disabled).
```
Comment (reyang) [inline src/OpenTelemetry/SuppressInstrumentationScope.cs:38]: Do we want to remove "automatic"?

```suggestion
        /// Begins a new scope in which instrumentation is suppressed (disabled).
```

File: src/OpenTelemetry/SuppressInstrumentationScope.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        /// </summary>
+        /// <param name="value">Value indicating whether to suppress automatic telemetry.</param>
```
Comment (reyang) [inline src/OpenTelemetry/SuppressInstrumentationScope.cs:40]: Similar here, wonder if we should change the wording.

File: src/OpenTelemetry/SuppressInstrumentationScope.cs (hunk 4 above)
Comment (alanwest) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:35]: 😄 agree `unused` is better than `x`. I changed it. Though this did leave me wondering if people would prefer to be able to use `_` in this case. I think it's pretty common practice, though I imagine allowing it means allowing any variable to start with a `_` which I'm guessing we don't want. No biggie `unused` is great.

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs]: I like `_` and initially I was trying to use it, then caught by a style cop 😄

**Comment by cijothomas** [comment]
@alanwest  conflict here :-)

**Comment by alanwest** [comment]
You're a mergin' machine, @cijothomas! Thanks for shepherding so many PRs through!

**Pull request merged**